### ISSUE TITLE
Any up to date document or tutorials?

### ISSUE THREAD

**Issue by asesidaa** [gmu-swe/phosphor#168]
Hi, we are trying to use Phosphor as part of our project.

When we try to instrument the class with `MultiTainter.taintedX` method, we always get `java.lang.IllegalStateException: Calling uninstrumented Phosphor stubs!`, can you explain the life cycle of Phosphor?

Thank you

**Comment by jon-bell** [comment]
The README and our [tool demo](https://mice.cs.columbia.edu/getTechreport.php?techreportID=1601) should still be accurate here. The error message that you are receiving indicates that the code that is running was not instrumented by Phosphor. Please feel free to share the exact commands that you are running if you would like assistance debugging why this is happening.

**Comment by asesidaa** [comment]
So I figured out that the problem is that I write a wrapper around `MultiTainter.taintedX` (to add some logging and testing logic), and this wrapper is not processed by Phosphor because it is ignored.

I wonder what is the correct way to add a wrapper around `MultiTainter.taintedX`, since if I let Phospher process it, it will call a method tagged with `$$PHOSPHORTAGGED`, which does not exist in my code.

Thank you

**Comment by jon-bell** [comment]
*All* code must be instrumented by Phosphor, including your wrapper. Phosphor will instrument calls to your wrapper to adding the `$$PHOSPHORTAGGED`, and will also instrument the declaration of the method to add `$$PHOSPHORTAGGED`.

**Comment by asesidaa** [comment]
Now we can taint through instrumenting, but when we tried to taint static fields, the taint tags are always empty. I wonder why would this be the case?
```Java
import edu.columbia.cs.psl.phosphor.runtime.MultiTainter;
import edu.columbia.cs.psl.phosphor.runtime.Taint;
import org.junit.Test;

@SuppressWarnings("rawtypes")
public class DirectTaintTest {
    public static int[] a = MultiTainter.taintedIntArray(new int[3], "original1");

    public static String s = MultiTainter.taintedReference("123", "original2");

    @Test
    public void test1() {
        a[1] = MultiTainter.taintedInt(999, "label1");
        s = MultiTainter.taintedReference("456", "label2");
    }

    @Test
    public void test2(){
        Taint taintA = MultiTainter.getTaint(a[1]);
        // true, Empty
        System.out.println(taintA.isEmpty());
        System.out.printf("a[1] is %d label: %s%n", a[1], taintA);
        Taint taintS = MultiTainter.getTaint(s);
        // true, Empty
        System.out.println(taintS.isEmpty());
        System.out.printf("s is %s, label is %s%n", s, taintS);

        a[1] = MultiTainter.taintedInt(9999, "label3");
        s = MultiTainter.taintedReference("4567", "label4");
        taintA = MultiTainter.getTaint(a[1]);
        taintS = MultiTainter.getTaint(s);
        // label3
        System.out.printf("a[1] is %d label: %s%n", a[1], taintA);
        // label4
        System.out.printf("s is %s, label is %s%n", s, taintS);
    }
}
```

**Comment by katherine-hough** [comment]
When running your example, "test1" runs before "test2" printing the following output:
> false
a[1] is 999 label: Taint [Labels = [[original1, label1]]
false
s is 456, label is Taint [Labels = [[label2]]
a[1] is 9999 label: Taint [Labels = [[original1, label3]]
s is 4567, label is Taint [Labels = [[label4]]

I believe that this output is correct. The label "original1" propagates to taintA because the reference to the field "a" is tainted by the call to MultiTainter.taintedIntArray. When an element is accessed from an array, the resulting value's taint tag will be the union of the taint tag for the reference to the array, the taint tag for the index used to access that element, and the taint tag of the accessed element.

**Comment by asesidaa** [comment]
Hi, thank you for reply.

I don't seem to be able to get the same result, in my case, i will get empty taint for first `a[1]`.

I am using `0.0.5-SNAPSHOT` in my pom.xml, are there any other configurations that I need to use?

**Comment by katherine-hough** [comment]
Could you create a failing test in Phosphor or a minimal project reproducing your issue?

If you use the java option -DphosphorCacheDirectory=<PATH> where <PATH> is a path to some directory. Phosphor will cache instrumented classes in that directory. You could then examine the instrumented class file produced for DirectTaintTest using javap, the Java class file disassembler, by running the command "javap -private -verbose <CLASS_FILE>" where <CLASS_FILE> is the instrumented class file produced for DirectTaintTest. See https://docs.oracle.com/javase/7/docs/technotes/tools/windows/javap.html for more details on javap.

**Comment by asesidaa** [comment]
Hi Katherine, [Yiwei](https://github.com/victoryang00) has sent the invitation to our fork of Bramble to you. Our minimal reproducible example is in the origin branch. If you run it, you can see that the taint tag is not propagated across method calls, using the test case described above. We would like to find out what is causing this issue, Phosphor or Bramble, can you take a look and help us find it out?

Thank you

**Comment by jon-bell** [comment]
We are not able to provide significant support for Bramble. If you are unable to create a minimum reproducible example outside of Bramble, then the issue is there. Please see if you can provide us with a minimum reproducible example here.

**Comment by asesidaa** [comment]
Hi, I have tried to run Phosphor directly from command line, but the result is correct...

Therefore I tried to compare the difference. I have tried to compare the instrumented byte code for the test case. However, aside from the the order of local variable creation, the instrumented byte code files are identical, including the variable index. (The two versions of instrumented byte code are attached below) 

I also tried to keep the arguments to phosphor, classpath and the instrumented jre the same, but the result is still different. Are there any other possible cause for the different result?

Thank you
[testcase.zip](https://github.com/gmu-swe/phosphor/files/8861384/testcase.zip)
